src/Core/src/HttpClient/RetryResponse.php

```diff
@@ -0,0 +1,208 @@
+    {
+        $this->client = $client;
+        $this->method = $method;
+        $this->url = $url;
+        $this->options = $options;
+        $this->maxTryCount = $maxTryCount;
+
+        $this->logger = $logger ?? new NullLogger();
+        $this->inner = $this->client->request($this->method, $this->url, $this->options);
+    }
+
+    public function getStatusCode(): int
+    {
+        $this->initialized or $this->initialize();
+
+        return $this->inner->getStatusCode();
+    }
+
+    public function getHeaders(bool $throw = true): array
+    {
+        $this->initialized or $this->initialize();
+
+        return $this->inner->getHeaders($throw);
+    }
+
+    public function getContent(bool $throw = true): string
+    {
+        $this->initialized or $this->initialize();
+
+        return $this->inner->getContent($throw);
+    }
+
+    public function toArray(bool $throw = true): array
+    {
+        $this->initialized or $this->initialize();
+
+        return $this->inner->toArray($throw);
+    }
+
+    public function cancel(): void
+    {
+        $this->initialized = true;
+        $this->inner->cancel();
+    }
+
+    public function getInfo(string $type = null)
+    {
+        $this->initialized or $this->initialize();
+
+        return $this->inner->getInfo($type);
+    }
+
+    public static function stream(iterable $responses, float $timeout = null): \Generator
+    {
+        $wrappedResponses = [];
+        $map = new \SplObjectStorage();
+        $client = null;
+
+        foreach ($responses as $r) {
+            if (!$r instanceof self) {
+                throw new \TypeError(sprintf('"%s::stream()" expects parameter 1 to be an iterable of AsyncResponse objects, "%s" given.', static::class, get_debug_type($r)));
+            }
+
+            if (null === $client) {
+                $client = $r->client;
+            } elseif ($r->client !== $client) {
+                throw new TransportException('Cannot stream AsyncResponse objects with many clients.');
+            }
+        }
+
+        if (!$client) {
+            return;
+        }
+
+        $end = null;
+        if (null !== $timeout) {
+            $end = \microtime(true) + $timeout;
+        }
+
+        while (true) {
+            foreach ($responses as $r) {
+                $wrappedResponses[] = $r->inner;
+                $map[$r->inner] = $r;
+            }
+            $subTimeout = null;
+            if (null !== $end && ($subTimeout = $end - \microtime(true)) <= 0) {
+                foreach ($map as $response) {
+                    yield $response => new ErrorChunk(0, sprintf('Idle timeout reached for "%s".', $response->getInfo('url')));
+                }
+
+                break;
+            }
+
+            foreach ($client->stream($wrappedResponses, $subTimeout) as $response => $chunk) {
+                /** @var self $r */
+                $r = $map[$response];
+                if (!$chunk->isTimeout() && $chunk->isFirst()) {
```

Comment:
I don’t understand this. 
If it isn’t a retry and the first chunk...

Reply:
the method `handleRetry` checks if a retry is needed and possible, and return true id the request is retried.

we only check for a retry on first chunk, because we only need the status code to decide if we need to retry the request.

Reply:
But if it is a timeout, dont we want to retry then?

Reply:
I think `timeout` here occurred when you stream the response with a userland timeout. The request is not resolved at this point. 
A timeout in network communication would throw a `NetworkException`.

Am I right @nicolas-grekas ?